src/crimson/tools/store_bench/store-bench.cc

```diff
@@ -0,0 +1,685 @@
+        tot_ops_all_io);
+  ERROR("Total latency aka time per operation, across ios is {}",
+        tot_latency_all_io);
+  ERROR("throughput, number of ops per millisec is {}",
+        static_cast<double>(tot_ops_all_io) / duration);
+  ERROR("average latency per io is  {}",
+        tot_latency_all_io / tot_ops_all_io);
+  co_return;
+};
+
+/**
+ * This function adds and removes log entries to a log object
+ * It returns throughput(number of operations/milli sec)
+ * Also returns average latency (time/operation)
+ * This function is split into 3 steps:
+ * (a)pre filling the logs to steady state
+ * (b) writing and removing logs ,this is considered 1 I/O
+ * (c) doing N I/o's concurrently on 1 thread
+ */
+seastar::future<> pg_log_workload(crimson::os::FuturizedStore &global_store,
+                                  int num_logs, int num_concurrent_io,
+                                  int duration, int log_size, int log_length) {
+  LOG_PREFIX(pg_log_workload);
+  auto &local_store = global_store.get_sharded_store();
+
+  std::map<int, coll_t> collection_id;
+  std::map<int, boost::intrusive_ptr<crimson::os::FuturizedCollection>>
+      coll_ref_map;
+
+  /**
+   * This method returns a future with pre filled logs
+   * It creates num_log number of log objects
+   * Each log object has log_length number of entries
+   * Each entry is a key value pair, with the value having size log_size
+   */
+
+  auto pre_fill_logs = [&]() -> seastar::future<> {
+    for (int i = 0; i < num_logs; ++i) {
+      auto obj_i = create_hobj(i);
+      auto coll_id = make_cid(i);
+      collection_id[i] = coll_id;
+      auto coll_ref = co_await local_store.create_new_collection(coll_id);
+      coll_ref_map[i] = coll_ref;
+      std::map<std::string, bufferlist> data;
+      for (int j = 0; j < log_length; ++j) {
+        std::string key = std::to_string(j);
+        bufferlist bl_value;
+        bl_value.append_zero(log_size);
+        data[key] = bl_value;
+      }
+      ceph::os::Transaction txn;
+      txn.create(coll_id, obj_i);
+      txn.omap_setkeys(coll_id, obj_i, data);
+      co_await local_store.do_transaction(coll_ref, std::move(txn));
+    }
+    co_return;
+  };
+
+  std::vector<int> first_key_per_log(num_logs,
+                                     0); // first key in each log object
+  std::vector<int> last_key_per_log(num_logs,
+                                    log_length); // last key in each log object
+
+  /**
+   * This method returns a future of type struct results_t
+   * In this function we choose a random log object to write to and remove keys
+   * from We add and remove keys sequentially, add keys to the end, remove from
+   * the front Total latency per io is calculated as the sum of the time it
+   * takes to add and remove each key
+   */
+  auto add_remove_entry = [&]() -> seastar::future<results_t> {
+    int num_ops = 0;
+    int tot_latency = 0;
+    auto start = seastar::lowres_clock::now();
```

Comment:
This is a reasonable choice, but better to use ceph::mono_clock from common/ceph_time.h.